agent/engine/task_manager.go

```diff
@@ -159,31 +138,110 @@ func (mtask *managedTask) overseeTask() {
 		if err != nil {
 			llog.Warn("Error checkpointing task's states to disk", "err", err)
 		}
+
 		if mtask.GetKnownStatus().Terminal() {
 			break
 		}
 	}
 	// We only break out of the above if this task is known to be stopped. Do
 	// onetime cleanup here, including removing the task after a timeout
 	llog.Debug("Task has reached stopped. We're just waiting and removing containers now")
-	taskCredentialsID := mtask.GetCredentialsID()
-	if taskCredentialsID != "" {
-		mtask.engine.credentialsManager.RemoveCredentials(taskCredentialsID)
-	}
+	mtask.cleanupCredentials()
 	if mtask.StopSequenceNumber != 0 {
 		llog.Debug("Marking done for this sequence", "seqnum", mtask.StopSequenceNumber)
 		mtask.engine.taskStopGroup.Done(mtask.StopSequenceNumber)
 	}
 	mtask.cleanupTask(mtask.engine.cfg.TaskCleanupWaitDuration)
 }
 
+// emitCurrentStatus emits a container event for every container and a task
+// event for the task
 func (mtask *managedTask) emitCurrentStatus() {
 	for _, container := range mtask.Containers {
 		mtask.engine.emitContainerEvent(mtask.Task, container, "")
 	}
 	mtask.engine.emitTaskEvent(mtask.Task, "")
 }
 
+// waitForResources waits for host resources to become available to start the
+// task. This involves waiting for previous stops to complete so the resources
+// become free.
+func (mtask *managedTask) waitForResources() {
+	llog := log.New("task", mtask.Task)
+	if mtask.StartSequenceNumber != 0 && !mtask.GetDesiredStatus().Terminal() {
+		llog.Debug("Waiting for any previous stops to complete", "seqnum", mtask.StartSequenceNumber)
+		othersStopped := make(chan bool, 1)
+		go func() {
+			mtask.engine.taskStopGroup.Wait(mtask.StartSequenceNumber)
+			othersStopped <- true
+		}()
+		for !mtask.waitEvent(othersStopped) {
+			if mtask.GetDesiredStatus().Terminal() {
+				// If we end up here, that means we received a start then stop for this
+				// task before a task that was expected to stop before it could
+				// actually stop
+				break
+			}
+		}
+		llog.Debug("Wait over; ready to move towards status: " + mtask.GetDesiredStatus().String())
+	}
+	return
+}
+
+// waitSteady waits for a task to leave steady-state by waiting for a new
+// event, or a timeout.
+func (mtask *managedTask) waitSteady() {
+	llog := log.New("task", mtask.Task)
+	llog.Debug("Task at steady state", "state", mtask.GetKnownStatus().String())
+
+	maxWait := make(chan bool, 1)
+	timer := mtask.time().After(steadyStateTaskVerifyInterval)
+	go func() {
+		<-timer
+		maxWait <- true
```

Comment:
It isn't exactly a common idiom.  Part of the complication is that `waitEvent` is actually doing a `select` on three different channels in order to see if either an event occurred (a message from ACS or from Docker) or if it's canceled. If canceled, it returns the `bool` that was passed through the channel; otherwise it returns `false`.

I looked at moving this over to `context.Context` entirely and using a deadline, but that ended up being more work than I wanted to put in (we have some tests that simulate a timeout and I'd have to change those too).